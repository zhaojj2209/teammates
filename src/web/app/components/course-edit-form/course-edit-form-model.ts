import { Course } from '../../../types/api-output';

/**
 * The mode of operation for question edit form.
 */
export enum CourseEditFormMode {
    /**
     * Adding a new course.
     */
    ADD,

    /**
     * Editing the existing course.
     */
    EDIT,
}

/**
 * Timezone format used along with course.
 */
export interface Timezone {
    id: string;
    offset: string;
}

/**
 * The form model of course form.
 */
export interface CourseFormModel {
    course: Course;
    timezones: Timezone[];

    isSaving: boolean;
}

/**
 * The form model of course edit form in ADD mode.
 */
export interface CourseAddFormModel extends CourseFormModel {
    institutes: string[];
    activeCourses: Course[];
    allCourses: Course[];

    isCopying: boolean;
}

/**
 * The form model of course edit form in Edit mode.
 */
export interface CourseEditFormModel extends CourseFormModel {
    originalCourse: Course;

    isEditing: boolean;
    canModifyCourse: boolean;
}

const DefaultCourse: Course = {
    courseName: '',
    courseId: '',
    institute: '',
    timeZone: 'UTC',
    creationTimestamp: 0,
    deletionTimestamp: 0,
};

const DefaultCourseModel: CourseFormModel = {
    course: JSON.parse(JSON.stringify(DefaultCourse)),
    timezones: [],
    isSaving: false,
};

const DefaultCourseAddFormModel: CourseAddFormModel = {
    course: JSON.parse(JSON.stringify(DefaultCourse)),
    timezones: [],
    institutes: [],
    activeCourses: [],
    allCourses: [],

    isSaving: false,
    isCopying: false,
};

const DefaultCourseEditFormModel: CourseEditFormModel = {
    course: JSON.parse(JSON.stringify(DefaultCourse)),
    originalCourse: JSON.parse(JSON.stringify(DefaultCourse)),
    timezones: [],

    isSaving: false,
    isEditing: false,
    canModifyCourse: false,
};

/**
 * Returns default course form model.
 */
export const DEFAULT_COURSE_FORM_MODEL: Function = (): CourseEditFormModel => {
    return JSON.parse(JSON.stringify(DefaultCourseModel));
};

/**
 * Returns default course edit form model.
 */
export const DEFAULT_COURSE_EDIT_FORM_MODEL: Function = (): CourseEditFormModel => {
    return JSON.parse(JSON.stringify(DefaultCourseEditFormModel));
};

/**
 * Returns default course add form model.
 */
export const DEFAULT_COURSE_ADD_FORM_MODEL: Function = (): CourseAddFormModel => {
    return JSON.parse(JSON.stringify(DefaultCourseAddFormModel));
};
