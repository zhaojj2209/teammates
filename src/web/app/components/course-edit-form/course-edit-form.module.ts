import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NgbTooltipModule } from '@ng-bootstrap/ng-bootstrap';
import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
import { LoadingRetryModule } from '../loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../loading-spinner/loading-spinner.module';
import { CourseEditFormComponent } from './course-edit-form.component';

@NgModule({
  declarations: [
    CourseEditFormComponent,
  ],
  imports: [
    CommonModule,
    FormsModule,
    AjaxLoadingModule,
    NgbTooltipModule,
    LoadingRetryModule,
    LoadingSpinnerModule,
  ],
  exports: [
    CourseEditFormComponent,
  ],
})
export class CourseEditFormModule { }
