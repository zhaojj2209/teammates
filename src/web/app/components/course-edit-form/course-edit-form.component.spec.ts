import { CommonModule } from '@angular/common';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { EventEmitter } from '@angular/core';
import {
 ComponentFixture,
 TestBed,
 waitForAsync,
} from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModal, NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { of } from 'rxjs';

import { CourseService } from '../../../services/course.service';
import { FeedbackSessionsService } from '../../../services/feedback-sessions.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { TimezoneService } from '../../../services/timezone.service';
import { createMockNgbModalRef } from '../../../test-helpers/mock-ngb-modal-ref';
import {
  Course,
  FeedbackSession,
  FeedbackSessionPublishStatus,
  FeedbackSessionSubmissionStatus,
  ResponseVisibleSetting,
  SessionVisibleSetting,
} from '../../../types/api-output';
import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
import { CopyCourseModalComponent } from '../copy-course-modal/copy-course-modal.component';
import { LoadingRetryModule } from '../loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../loading-spinner/loading-spinner.module';
import {
  CourseAddFormModel,
   CourseEditFormMode,
   DEFAULT_COURSE_ADD_FORM_MODEL,
   DEFAULT_COURSE_EDIT_FORM_MODEL,
} from './course-edit-form-model';
import { CourseEditFormComponent } from './course-edit-form.component';

describe('CourseEditFormComponent', () => {
  let component: CourseEditFormComponent;
  let fixture: ComponentFixture<CourseEditFormComponent>;
  let feedbackSessionsService: FeedbackSessionsService;
  let ngbModal: NgbModal;

  const validCourseId: string = 'CS1101S';
  const validTimeZone: string = 'Asia/Singapore';
  const validInstitute: string = 'Test Institute';
  const testTimeZone: string = 'Australia/Adelaide';
  const timeZoneOffsets1: Record<string, number> = { GMT: 570 };
  const testCourse: Course = {
    courseId: 'testId',
    courseName: 'Test Course',
    timeZone: validTimeZone,
    institute: validInstitute,
    creationTimestamp: 0,
    deletionTimestamp: 1000,
  };

  const spyStatusMessageService: any = {
    showErrorToast: jest.fn(),
    showSuccessToast: jest.fn(),
  };
  const timezoneServiceStub: any = {
    getTzOffsets: jest.fn(() => timeZoneOffsets1),
    guessTimezone: jest.fn(() => testTimeZone),
  };

  const spyCourseService: any = {
    createCourse: jest.fn(() => of({})),
    getAllCoursesAsInstructor: jest.fn(() => of({ courses: [testCourse] })),
  };

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [CourseEditFormComponent],
      imports: [
        HttpClientTestingModule,
        FormsModule,
        ReactiveFormsModule,
        RouterTestingModule,
        NgbModule,
        AjaxLoadingModule,
        CommonModule,
        FormsModule,
        AjaxLoadingModule,
        LoadingRetryModule,
        LoadingSpinnerModule,
      ],
      providers: [
        { provide: StatusMessageService, useValue: spyStatusMessageService },
        { provide: CourseService, useValue: spyCourseService },
        { provide: TimezoneService, useValue: timezoneServiceStub },
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CourseEditFormComponent);
    feedbackSessionsService = TestBed.inject(FeedbackSessionsService);
    ngbModal = TestBed.inject(NgbModal);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap when in ADD Mode and is display only', () => {
    component.isDisplayOnly = true;
    component.formMode = CourseEditFormMode.ADD;
    fixture.detectChanges();

    const optionElement = fixture.debugElement.nativeElement.querySelector('#time-zone');
    const submitButton = fixture.debugElement.nativeElement.querySelector('#btn-submit-course');
    const copyButton = fixture.debugElement.nativeElement.querySelector('#btn-copy-course');

    expect(submitButton).toHaveProperty('disabled', true);
    expect(copyButton).toHaveProperty('disabled', true);
    expect(optionElement.options.length).toEqual(0);
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when in ADD Mode with default fields', () => {
    component.isDisplayOnly = false;
    component.formMode = CourseEditFormMode.ADD;
    component.formModel = DEFAULT_COURSE_ADD_FORM_MODEL();
    fixture.detectChanges();

    const optionElement = fixture.debugElement.nativeElement.querySelector('#time-zone');
    const submitButton = fixture.debugElement.nativeElement.querySelector('#btn-submit-course');
    const copyButton = fixture.debugElement.nativeElement.querySelector('#btn-copy-course');

    // submit button and copy button must be disabled since there is no valid input for the input fields
    expect(submitButton).toHaveProperty('disabled', true);
    expect(copyButton).toHaveProperty('disabled', true);

    expect(optionElement.options.length).toEqual(1);
    expect(component.model.course.timeZone).toEqual(testTimeZone);
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when in EDIT Mode and is display only', () => {
    component.isDisplayOnly = true;
    component.formMode = CourseEditFormMode.EDIT;
    fixture.detectChanges();

    const optionElement = fixture.debugElement.nativeElement.querySelector('#time-zone');
    const deleteButton = fixture.debugElement.nativeElement.querySelector('#btn-delete-course');
    const editButton = fixture.debugElement.nativeElement.querySelector('#btn-edit-course');

    expect(optionElement.options.length).toEqual(0);
    expect(deleteButton).toHaveProperty('disabled', true);
    expect(editButton).toHaveProperty('disabled', true);
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when in EDIT Mode with default fields', () => {
    component.isDisplayOnly = false;
    component.formMode = CourseEditFormMode.EDIT;
    component.formModel = DEFAULT_COURSE_EDIT_FORM_MODEL();
    fixture.detectChanges();

    const deleteButton = fixture.debugElement.nativeElement.querySelector('#btn-delete-course');
    const editButton = fixture.debugElement.nativeElement.querySelector('#btn-edit-course');

    expect(deleteButton).toHaveProperty('disabled', true);
    expect(editButton).toHaveProperty('disabled', true);
    expect(fixture).toMatchSnapshot();
  });

  it('should open copy course modal when in ADD Mode', () => {
    component.formModel = DEFAULT_COURSE_ADD_FORM_MODEL();
    component.formMode = CourseEditFormMode.ADD;
    fixture.detectChanges();

    const mockModalRef: any = createMockNgbModalRef({
      isCopyFromOtherSession: false,
      courses: [],
      courseToFeedbackSession: {},
      fetchFeedbackSessionsEvent: new EventEmitter<string>(),
    });

    const testFeedbackSession: FeedbackSession = {
      courseId: validCourseId,
      timeZone: validTimeZone,
      feedbackSessionName: 'Test Session',
      instructions: 'Instructions',
      submissionStartTimestamp: 1000000000000,
      submissionEndTimestamp: 1500000000000,
      gracePeriod: 0,
      sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
      responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
      submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
      publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
      isClosingEmailEnabled: true,
      isPublishedEmailEnabled: true,
      createdAtTimestamp: 0,
      studentDeadlines: {},
      instructorDeadlines: {},
    };

    jest.spyOn(feedbackSessionsService, 'getFeedbackSessionsForInstructor')
      .mockReturnValue(of({ feedbackSessions: [testFeedbackSession] }));
    jest.spyOn(ngbModal, 'open').mockReturnValue(mockModalRef);

    const model: CourseAddFormModel = component.model as CourseAddFormModel;
    model.activeCourses = [testCourse];

    component.copyCourseHandler();
    mockModalRef.componentInstance.fetchFeedbackSessionsEvent.emit(testCourse.courseId);

    expect(ngbModal.open).toHaveBeenCalledWith(CopyCourseModalComponent);
    expect(mockModalRef.componentInstance.isCopyFromOtherSession).toEqual(true);
    expect(mockModalRef.componentInstance.activeCourses[0]).toEqual(testCourse);
    expect(mockModalRef.componentInstance.courseToFeedbackSession[testCourse.courseId]).toEqual([testFeedbackSession]);
  });
});
