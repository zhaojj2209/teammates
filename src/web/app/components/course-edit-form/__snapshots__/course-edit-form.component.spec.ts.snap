// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseEditFormComponent should snap when in ADD Mode and is display only 1`] = `
<tm-course-edit-form
  CourseEditFormMode={[Function Object]}
  FormValidator={[Function Object]}
  __ngContext__={[Function Number]}
  addModel="undefined"
  closeFormEvent={[Function EventEmitter_]}
  copyCourseEvent={[Function EventEmitter_]}
  createNewCourseEvent={[Function EventEmitter_]}
  deleteCourseEvent={[Function EventEmitter_]}
  editModel="undefined"
  feedbackSessionsService={[Function FeedbackSessionsService]}
  form={[Function NgForm]}
  formMode="0"
  isDisplayOnly={[Function Boolean]}
  model={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetEventSubscription={[Function Subscription]}
  resetFormEvent={[Function EventEmitter_]}
  statusMessageService={[Function Object]}
  timezoneService={[Function Object]}
  updateCourseEvent={[Function EventEmitter_]}
>
  <div
    class="card"
  >
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        Create a new Course
      </strong>
      <div
        class="card-header-btn-toolbar"
      >
        <button
          class="btn btn-primary btn-sm"
          disabled=""
          id="btn-copy-course"
        >
          <i
            aria-hidden="true"
            class="fa fa-copy"
          />
           Copy from previous courses 
        </button>
        <button
          class="btn btn-primary btn-sm"
          id="btn-close-form"
        >
          <i
            aria-hidden="true"
            class="fa fa-times"
          />
        </button>
      </div>
    </div>
    <div
      class="card-body fill-plain"
    >
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-id"
          >
            Course ID:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="col-xs-12 form-control ng-untouched ng-pristine ng-valid"
              id="course-id"
              maxlength="64"
              name="courseId"
              placeholder="e.g. CS3215-2013Semester1"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course ID should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-name"
          >
            Course Name:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-valid"
              id="course-name"
              maxlength="80"
              name="courseName"
              placeholder="e.g. Software Engineering"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course Name should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-institute"
          >
            <span
              class="ngb-tooltip-class"
              ngbtooltip="Note that you can only create a new course under an institute in which you are already a course co-owner."
            >
               Course institute: 
            </span>
          </label>
          <div
            class="col-sm-9"
          >
            <select
              class="col-xs-12 form-control form-select ng-untouched ng-pristine ng-valid"
              id="course-institute"
              name="institute"
              required=""
              type="text"
            >
            </select>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-xs-12 col-sm-2 fw-bold text-md-end"
            for="time-zone"
          >
            <span
              class="ngb-tooltip-class"
              ngbtooltip="You should not need to change this as it is auto-detected based on your
                  device settings. TEAMMATES automatically adjusts to match the current time offset in your area, including clock changes due to daylight saving time."
            >
               Time Zone:
            </span>
          </label>
          <div
            class="col-xs-12 col-sm-9"
          >
            <div
              class="input-group"
            >
              <select
                class="form-control form-select ng-untouched ng-pristine ng-valid"
                id="time-zone"
                name="timeZone"
                required=""
              >
              </select>
              <span
                class="input-group-btn"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="button"
                >
                   Auto-Detect 
                </button>
              </span>
            </div>
          </div>
        </div>
        <div
          class="form-group m-0"
        >
          <div
            class="col-sm-12 text-center"
          >
            <button
              class="btn btn-success"
              disabled=""
              id="btn-submit-course"
              type="submit"
            >
               Add Course 
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</tm-course-edit-form>
`;

exports[`CourseEditFormComponent should snap when in ADD Mode with default fields 1`] = `
<tm-course-edit-form
  CourseEditFormMode={[Function Object]}
  FormValidator={[Function Object]}
  __ngContext__={[Function Number]}
  addModel={[Function Object]}
  closeFormEvent={[Function EventEmitter_]}
  copyCourseEvent={[Function EventEmitter_]}
  createNewCourseEvent={[Function EventEmitter_]}
  deleteCourseEvent={[Function EventEmitter_]}
  editModel="undefined"
  feedbackSessionsService={[Function FeedbackSessionsService]}
  form={[Function NgForm]}
  formMode="0"
  isDisplayOnly="false"
  model={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetEventSubscription={[Function SafeSubscriber]}
  resetFormEvent={[Function EventEmitter_]}
  statusMessageService={[Function Object]}
  timezoneService={[Function Object]}
  updateCourseEvent={[Function EventEmitter_]}
>
  <div
    class="card"
  >
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        Create a new Course
      </strong>
      <div
        class="card-header-btn-toolbar"
      >
        <button
          class="btn btn-primary btn-sm"
          disabled=""
          id="btn-copy-course"
        >
          <i
            aria-hidden="true"
            class="fa fa-copy"
          />
           Copy from previous courses 
        </button>
        <button
          class="btn btn-primary btn-sm"
          id="btn-close-form"
        >
          <i
            aria-hidden="true"
            class="fa fa-times"
          />
        </button>
      </div>
    </div>
    <div
      class="card-body fill-plain"
    >
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-id"
          >
            Course ID:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="col-xs-12 form-control ng-untouched ng-pristine ng-valid"
              id="course-id"
              maxlength="64"
              name="courseId"
              placeholder="e.g. CS3215-2013Semester1"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course ID should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-name"
          >
            Course Name:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-valid"
              id="course-name"
              maxlength="80"
              name="courseName"
              placeholder="e.g. Software Engineering"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course Name should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-institute"
          >
            <span
              class="ngb-tooltip-class"
              ngbtooltip="Note that you can only create a new course under an institute in which you are already a course co-owner."
            >
               Course institute: 
            </span>
          </label>
          <div
            class="col-sm-9"
          >
            <select
              class="col-xs-12 form-control form-select ng-untouched ng-pristine ng-valid"
              id="course-institute"
              name="institute"
              required=""
              type="text"
            >
            </select>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-xs-12 col-sm-2 fw-bold text-md-end"
            for="time-zone"
          >
            <span
              class="ngb-tooltip-class"
              ngbtooltip="You should not need to change this as it is auto-detected based on your
                  device settings. TEAMMATES automatically adjusts to match the current time offset in your area, including clock changes due to daylight saving time."
            >
               Time Zone:
            </span>
          </label>
          <div
            class="col-xs-12 col-sm-9"
          >
            <div
              class="input-group"
            >
              <select
                class="form-control form-select ng-untouched ng-pristine ng-valid"
                id="time-zone"
                name="timeZone"
                required=""
              >
                <option
                  value="GMT"
                >
                  GMT (UTC +09:30)
                </option>
              </select>
              <span
                class="input-group-btn"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                   Auto-Detect 
                </button>
              </span>
            </div>
          </div>
        </div>
        <div
          class="form-group m-0"
        >
          <div
            class="col-sm-12 text-center"
          >
            <button
              class="btn btn-success"
              disabled=""
              id="btn-submit-course"
              type="submit"
            >
               Add Course 
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</tm-course-edit-form>
`;

exports[`CourseEditFormComponent should snap when in EDIT Mode and is display only 1`] = `
<tm-course-edit-form
  CourseEditFormMode={[Function Object]}
  FormValidator={[Function Object]}
  __ngContext__={[Function Number]}
  addModel="undefined"
  closeFormEvent={[Function EventEmitter_]}
  copyCourseEvent={[Function EventEmitter_]}
  createNewCourseEvent={[Function EventEmitter_]}
  deleteCourseEvent={[Function EventEmitter_]}
  editModel="undefined"
  feedbackSessionsService={[Function FeedbackSessionsService]}
  form={[Function NgForm]}
  formMode={[Function Number]}
  isDisplayOnly={[Function Boolean]}
  model={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetEventSubscription={[Function Subscription]}
  resetFormEvent={[Function EventEmitter_]}
  statusMessageService={[Function Object]}
  timezoneService={[Function Object]}
  updateCourseEvent={[Function EventEmitter_]}
>
  <div
    class="card"
  >
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        Course
      </strong>
      <div
        class="card-header-btn-toolbar"
      >
        <button
          class="btn btn-primary btn-sm"
          disabled=""
          id="btn-edit-course"
        >
          <i
            aria-hidden="true"
            class="fa fa-pen"
          />
           Edit 
        </button>
        <button
          class="btn btn-primary btn-sm"
          disabled=""
          id="btn-delete-course"
          ngbtooltip="Delete the course and its corresponding students and sessions"
        >
          <i
            aria-hidden="true"
            class="fa fa-trash"
          />
           Delete 
        </button>
      </div>
    </div>
    <div
      class="card-body fill-plain"
    >
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-id"
          >
            Course ID:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="col-xs-12 form-control ng-untouched ng-pristine ng-valid"
              id="course-id"
              maxlength="64"
              name="courseId"
              placeholder="e.g. CS3215-2013Semester1"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course ID should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-name"
          >
            Course Name:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-valid"
              id="course-name"
              maxlength="80"
              name="courseName"
              placeholder="e.g. Software Engineering"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course Name should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-institute"
          >
             Course institute: 
          </label>
          <div
            class="col-sm-9"
          >
            <select
              class="col-xs-12 form-control form-select ng-untouched ng-pristine ng-valid"
              id="course-institute"
              name="institute"
              required=""
              type="text"
            >
              <option
                selected=""
                value=""
              >
              </option>
            </select>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-xs-12 col-sm-2 fw-bold text-md-end"
            for="time-zone"
          >
            <span
              class="ngb-tooltip-class"
              ngbtooltip="You should not need to change this as it is auto-detected based on your
                  device settings. TEAMMATES automatically adjusts to match the current time offset in your area, including clock changes due to daylight saving time."
            >
               Time Zone:
            </span>
          </label>
          <div
            class="col-xs-12 col-sm-9"
          >
            <div
              class="input-group"
            >
              <select
                class="form-control form-select ng-untouched ng-pristine ng-valid"
                id="time-zone"
                name="timeZone"
                required=""
              >
              </select>
              <span
                class="input-group-btn"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="button"
                >
                   Auto-Detect 
                </button>
              </span>
            </div>
          </div>
        </div>
        <div
          class="form-group m-0"
        >
          <div
            class="col-sm-12 text-center"
          >
          </div>
        </div>
      </form>
    </div>
  </div>
</tm-course-edit-form>
`;

exports[`CourseEditFormComponent should snap when in EDIT Mode with default fields 1`] = `
<tm-course-edit-form
  CourseEditFormMode={[Function Object]}
  FormValidator={[Function Object]}
  __ngContext__={[Function Number]}
  addModel="undefined"
  closeFormEvent={[Function EventEmitter_]}
  copyCourseEvent={[Function EventEmitter_]}
  createNewCourseEvent={[Function EventEmitter_]}
  deleteCourseEvent={[Function EventEmitter_]}
  editModel={[Function Object]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  form={[Function NgForm]}
  formMode={[Function Number]}
  isDisplayOnly="false"
  model={[Function Object]}
  ngbModal={[Function NgbModal]}
  resetEventSubscription={[Function SafeSubscriber]}
  resetFormEvent={[Function EventEmitter_]}
  statusMessageService={[Function Object]}
  timezoneService={[Function Object]}
  updateCourseEvent={[Function EventEmitter_]}
>
  <div
    class="card"
  >
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        Course
      </strong>
      <div
        class="card-header-btn-toolbar"
      >
        <button
          class="btn btn-primary btn-sm"
          disabled=""
          id="btn-edit-course"
        >
          <i
            aria-hidden="true"
            class="fa fa-pen"
          />
           Edit 
        </button>
        <button
          class="btn btn-primary btn-sm"
          disabled=""
          id="btn-delete-course"
          ngbtooltip="Delete the course and its corresponding students and sessions"
        >
          <i
            aria-hidden="true"
            class="fa fa-trash"
          />
           Delete 
        </button>
      </div>
    </div>
    <div
      class="card-body fill-plain"
    >
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-id"
          >
            Course ID:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="col-xs-12 form-control ng-untouched ng-pristine ng-valid"
              id="course-id"
              maxlength="64"
              name="courseId"
              placeholder="e.g. CS3215-2013Semester1"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course ID should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-name"
          >
            Course Name:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-valid"
              id="course-name"
              maxlength="80"
              name="courseName"
              placeholder="e.g. Software Engineering"
              required=""
              type="text"
            />
            <div
              class="invalid-field"
              hidden=""
            >
              <i
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
               The field Course Name should not be empty. 
            </div>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-sm-2 fw-bold text-md-end"
            for="course-institute"
          >
             Course institute: 
          </label>
          <div
            class="col-sm-9"
          >
            <select
              class="col-xs-12 form-control form-select ng-untouched ng-pristine ng-valid"
              id="course-institute"
              name="institute"
              required=""
              type="text"
            >
              <option
                selected=""
                value=""
              >
              </option>
            </select>
          </div>
        </div>
        <div
          class="row form-group"
        >
          <label
            class="col-xs-12 col-sm-2 fw-bold text-md-end"
            for="time-zone"
          >
            <span
              class="ngb-tooltip-class"
              ngbtooltip="You should not need to change this as it is auto-detected based on your
                  device settings. TEAMMATES automatically adjusts to match the current time offset in your area, including clock changes due to daylight saving time."
            >
               Time Zone:
            </span>
          </label>
          <div
            class="col-xs-12 col-sm-9"
          >
            <div
              class="input-group"
            >
              <select
                class="form-control form-select ng-untouched ng-pristine ng-valid"
                id="time-zone"
                name="timeZone"
                required=""
              >
                <option
                  value="GMT"
                >
                  GMT (UTC +09:30)
                </option>
              </select>
              <span
                class="input-group-btn"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="button"
                >
                   Auto-Detect 
                </button>
              </span>
            </div>
          </div>
        </div>
        <div
          class="form-group m-0"
        >
          <div
            class="col-sm-12 text-center"
          >
          </div>
        </div>
      </form>
    </div>
  </div>
</tm-course-edit-form>
`;
